other/shared_company/src/lib.rs

```diff
@@ -0,0 +1,110 @@
+Shares can be bought (fixed rate for now).
+Funds can be spend using [Proposal] if accepted by a majority (of company_voting_tokens which are distributed with shares).
+This is only a demonstration. There are known security flaws in this design.
+(like: Buy a lot of shares, make proposal to send you all company_radix, vote for it).
+*/
+blueprint! {
+
+struct SharedCompany {
+    company_radix:  Vault,
+    company_shares: Vault,
+    company_voting_token: Vault,
+    share_counter: Decimal,
+    price_share: Decimal,
+    proposal_list: Vec<Address>,
+}
+
+
+impl SharedCompany {
+    // Creates a new component
+    pub fn new(price_share: Decimal) -> Component{
+
+
+        // shares are equity in the company. They can be bought for a fixed rate or returned for a part of the company vault
+        let shared_company_share_resource_def = ResourceBuilder::new()
+        .metadata("name", "SharedCompany share").metadata("symbol", "SC")
+        .new_token_fixed(1_000_000);
+
+        // voting tokens are used to vote on proposals
+        let shared_company_voting_token_resource_def = ResourceBuilder::new()
+        .metadata("name", "SharedCompany voting token").metadata("symbol", "SCVT")
+        .new_token_fixed(1_000_000);
+
+        // stores the addresses of proposal-components
+        let proposal_address_list: Vec<Address>= Vec::new();
+
+
+    //populate the SharedCompany struct and instantiate a new component
+    Self {
+        company_shares: Vault::with_bucket(shared_company_share_resource_def),
+        company_voting_token: Vault::with_bucket(shared_company_voting_token_resource_def),
+        company_radix: Vault::new(RADIX_TOKEN),
+        share_counter: Decimal(0.0 as i128),
+        price_share: price_share,
+        proposal_list: proposal_address_list
+    }.instantiate()
+
+    }
+
+    /// buys an amount of shares and returns change
+    pub fn buy_shares(&mut self, payment: Bucket) -> (Bucket, Bucket, Bucket) {
+        let max_share_buy_power = payment.amount() / self.price_share;
+        // Increase the share_counter so the amount of shares that are outstanding is tracked
+        self.share_counter += max_share_buy_power;
+        // take our price in XRD out of the payment and store it in the company vault
+        let collected_xrd = payment.take(self.price_share * max_share_buy_power);
+        self.company_radix.put(collected_xrd);
+        // return the share(s) and change
+        (self.company_shares.take(max_share_buy_power), payment, self.company_voting_token.take(max_share_buy_power))
+    }
+
+     /// sells an amount of shares and for a part of the companies xrd
+     pub fn sell_shares(&mut self, shares: Bucket, voting_token: Bucket) -> (Bucket, Bucket) {
+        if shares.amount() != voting_token.amount()  {return (shares, voting_token)};
+        // calculates the percentage of all shares
+        let percentage_of_all_shares = shares.amount() / self.share_counter;
```

Comment:
You don't have to keep track of the share count. You could just save the share's resource_def in the component's data. Then do something like `self.share_resource_def.supply() - self.company_share.amount()`